### DISCUSSION TITLE
Using Nats Jetstream as an event store

### DISCUSSION THREAD

**Discussion by cloudcompute** [synadia-labs/rita#8] [Q&A]
Hi @bruth

Any plans to work further on this (rita) library?

I have just asked a question [here](https://github.com/nats-io/nats-server/discussions/3772). Copying it here too.

a. Can we use NATS JetStream for Event Sourcing? Does it have the capability to store millions of events at a time without affecting the speed and performance of an application?

b. I read that we can archive the event logs to cold storage. But I have not come across an example that demonstrates how we can move the logs to, let's say, S3 storage. Is this "cold storage" feature as powerful as Pulsar's Tiered Storage?

c. Are there any applications who are using it as an Event Store in production? If yes, any feedback?

Thank you all for making such a great software and make it available to all of us.

**Comment by bruth** [comment]
I [responded](https://github.com/nats-io/nats-server/discussions/3772#discussioncomment-4627251) on your first post.

> Any plans to work further on this (rita) library?

Yes. There are a handful of abstractions and implementations I have in mind, but I appreciate any feedback. Requests and concrete use cases will motivate additional development.

Also for context, since I published this repo, I have joined Synadia, for whom the founder and CEO created NATS and employ the majority of maintainers. This is all to say, building out the the package/tooling to support ES is more relevant.

**Comment by cloudcompute** [comment]
Hi @bruth

Use cases for ES are many. Few of them are 

- B2B application where we can track inventory and movement of goods in real-time and make decisions instantly

- an application in which money is involved and replaying of existing commands/events is required in case any bug is found at a later point in time. For example, we made a mistake calculating taxes and we replay the TaxCalculated events and make the corrections. 

- For gaming..
https://streamnative.io/blog/case/2022-11-15-how-proxima-beta-implemented-cqrs-and-event-sourcing-on-top-of-apache-pulsar-and-scylladb/

I have found 3 libraries/frameworks that are used in production:

- Commanded written in Elixir - its fantastic feature is that it supports both Strong and Eventual consistency

- Axon and Eventuate frameworks written in Java 

The following links can really help someone to make a list of the features that should be there in an ES library.

Build Elixir CQRS/ES applications
https://github.com/commanded/commanded/

Real-world example backend implementing the CQRS/ES pattern in Elixir and Phoenix
https://github.com/slashdotdash/conduit

https://github.com/eventuate-local/eventuate-local
https://eventuate.io/gettingstarted-es.html

https://github.com/AxonFramework/AxonFramework

Companies using Commanded
https://github.com/commanded/commanded/wiki/Companies-using-Commanded

I think developers/companies won't mind buying a production-ready solution for ES/CQRS since it is quite a complex subject.

One more point. Functional programming is on the rise. If a solution exposes functional interfaces (examples), it'd be great.
